File: example/calc.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

#include "calculate.h"


int main(int argc, char *argv[]) {
    Expression expression;
    char variables[256];
    char output[4096];
    char errors[64];
    double *values;
    double result;
    int i;

    if (argc % 2 == 0) {
        if (argc == 2)
            strcpy(variables, "");
        else
            strcpy(variables, argv[2]);
        for (i = 4; i < argc; i += 2) {
            strcat(variables, ",");
            strcat(variables, argv[i]);
        }
        expression = Calculate.create(argv[1], variables, errors);

        if (expression) {
            values = (double *) malloc((argc - 2) / 2 * sizeof(double));
            for (i = 3; i < argc; i += 2)
                values[(i + 1) / 2 - 2] = strtod(argv[i], NULL);

            result = Calculate.evaluate(
                expression, values, (argc - 2) / 2, errors
            );

            if (!strcmp(errors, "")) {
                printf("Expression:\n");
                Calculate.expression(expression, output);
                printf("%s\n", output);
                printf("Variables:\n");
                Calculate.variables(expression, output);
                printf("%s\n", output);
                printf("Infix notation:\n");
                Calculate.infix(expression, output);
                printf("%s\n", output);
                printf("Postfix notation:\n");
                Calculate.postfix(expression, output);
                printf("%s\n", output);
                printf("Expression tree:\n");
                Calculate.tree(expression, output);
                printf("%s\n", output);
                printf("Result:\n");
                printf("%f\n", result);
            }
            else
                printf("%s\n", errors);
        }
        else {
            printf("%s\n", errors);
        }
        Calculate.free(expression);

    }

    return 0;
}

File: example/calc.py
#!/usr/bin/env python
from __future__ import absolute_import
from __future__ import print_function

import sys

from calculate import Expression
from calculate.exceptions import BaseCalculateException


if __name__ == '__main__':

    argc = len(sys.argv)
    if argc > 0 and argc % 2 == 0:

        try:
            variables = []
            values = []
            for i, arg in enumerate(sys.argv):
                if i < 2:
                    continue
                elif i % 2 == 0:
                    variables.append(arg)
                else:
                    values.append(float(arg))

            expression = Expression(sys.argv[1], variables)
            result = expression(values)

            print('Expression:')
            print(expression.expression())
            print('Variables:')
            print(','.join(expression.variables()))
            print('Infix notation:')
            print(expression.infix())
            print('Postfix notation:')
            print(expression.postfix())
            print('Expression tree:')
            print(expression.tree())
            print('Result:')
            print(result)

        except BaseCalculateException as e:
            print(e.message)
